File: src/class070/C3_MaximumSumCircularSubarray.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>
using namespace std;

class Code03_MaximumSumCircularSubarray {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int n = nums.size();
        int all = nums[0], maxSum = nums[0], minSum = nums[0];
        int maxPre = nums[0], minPre = nums[0];
        
        for (int i = 1; i < n; i++) {
            all += nums[i];
            maxPre = max(nums[i], nums[i] + maxPre);
            maxSum = max(maxSum, maxPre);
            minPre = min(nums[i], nums[i] + minPre);
            minSum = min(minSum, minPre);
        }

        // Handle the case where all elements are negative
        if (all == minSum) {
            return maxSum;
        }
        // Return the maximum of normal max subarray and circular subarray
        return max(maxSum, all - minSum);
    }
};

int main() {
    Code03_MaximumSumCircularSubarray solution;
    vector<int> nums = {5, -3, 5};
    cout << "Maximum Circular Subarray Sum: " << solution.maxSubarraySumCircular(nums) << endl;
    return 0;
}

File: src/class070/C3_MaximumSumCircularSubarray.py
class Code03_MaximumSumCircularSubarray:
    def max_subarray_sum_circular(self, nums):
        n = len(nums)
        all_sum = nums[0]
        max_sum = nums[0]
        min_sum = nums[0]
        max_pre = nums[0]
        min_pre = nums[0]

        for i in range(1, n):
            all_sum += nums[i]
            max_pre = max(nums[i], nums[i] + max_pre)
            max_sum = max(max_sum, max_pre)
            min_pre = min(nums[i], nums[i] + min_pre)
            min_sum = min(min_sum, min_pre)

        # Handle the case where all elements are negative
        if all_sum == min_sum:
            return max_sum
        # Return the maximum of normal max subarray and circular subarray
        return max(max_sum, all_sum - min_sum)

# Example usage
if __name__ == "__main__":
    solution = Code03_MaximumSumCircularSubarray()
    nums = [5, -3, 5]
    print("Maximum Circular Subarray Sum:", solution.max_subarray_sum_circular(nums))
